src/Juvix/FrontendContextualise/ModuleOpen/Environment.hs

```diff
@@ -100,3 +159,189 @@ lookupModMap s =
+
+--------------------------------------------------------------------------------
+-- fully resolve module opens
+--------------------------------------------------------------------------------
+
+-- Precondition ∷ old and new context must be in the same context
+-- also terms must be in either the old map or the new map
+-- any place where this is inconsistent will break resolution
+
+-- | @populateModMap@ populates the modMap with all the global opens
+-- in the module
+populateModMap ::
+  WorkingMaps m => m ()
+populateModMap = do
+  old <- get @"old"
+  new <- get @"new"
+  open <- Juvix.Library.ask @"openMap"
+  modM <- get @"modMap"
+  let curr = pure Context.topLevelName <> Context.currentName old
+  case open Map.!? curr of
+    Nothing ->
+      pure ()
+    Just opens ->
+      -- TODO ∷
+      -- explicts and implicits for now work the same
+      put @"modMap" (foldr f modM opens)
+      where
+        f y modMap = foldr addNewSymbol modMap openList
+          where
+            nameSpace =
+              case y of
+                Implicit x -> x
+                Explicit x -> x
+            unpopulate = (() <$)
+            --
+            addNewSymbol x modMap =
+              case unpopulate (Context.lookup (NameSymbol.fromSymbol x) old)
+                <|> unpopulate (Context.lookup (NameSymbol.fromSymbol x) new) of
+                Just () ->
+                  -- if the symbol is in the map, then
+                  -- we don't shadow it....
+                  -- TODO ∷ later throw an error for
+                  --        explicit opens that do this
+                  modMap
+                Nothing ->
+                  Map.insert x nameSpace modMap
+            --
+            openList =
+              -- haskell gives type error if grabList is not typed
+              let grabList ::
+                    Context.T a b c ->
+                    NameSpace.List (Context.Definition a b c)
+                  grabList ctx =
+                    case Context.extractValue <$> Context.lookup nameSpace ctx of
+                      Just (Context.Record nameSpace _) ->
+                        NameSpace.toList nameSpace
+                      Just _ ->
+                        NameSpace.List [] []
+                      Nothing ->
+                        NameSpace.List [] []
+                  NameSpace.List {publicL = pubN} =
+                    grabList new
+                  NameSpace.List {publicL = pubO} =
+                    grabList old
+               in fmap fst pubN <> fmap fst pubO
+
+-- we don't take a module mapping as all symbols at the point of
+-- resolve can't be anything else, thus we don't have to pre-pend any
+-- qualification at this point... we add qualification later as we add
+-- things, as we have to figure out the full resolution of all opens
+
+resolve :: Context.T a b c -> [PreQualified] -> Either Error OpenMap
+resolve ctx = foldM (resolveSingle ctx) mempty
+
+resolveSingle ::
```

Comment:
add a comment describing pre, post-conditions, function behaviour

Reply:
I'm not sure there are many preconditions for this function.

I think the `opens` part of the `PreQualified` has to not contain toplevel... but that is something fixable

Outside of that the ctx should be full by this point.

I guess that is a precondition

Reply:
This issue is now fixed, so there should be no preconditions other than ctx being realized by this point

Reply:
right, I guess I'm looking for an English description of what the function should do, that I can conceptualise and review the code against, otherwise it's tricky to know if it's correct or not